apps/avatax/src/modules/saleor/order.ts

```diff
@@ -47,3 +51,69 @@ export class SaleorOrder {
     return this.data.order.channel.taxConfiguration.pricesEnteredWithTax;
   }
 }
+
+interface ISaleorCancelledOrderEvent {
+  getChannelSlug(): string;
+  getAvataxId(): string;
+  getPrivateMetadata(): Array<{ key: string; value: string }>;
+}
+
+export class SaleorCancelledOrderEvent implements ISaleorCancelledOrderEvent {
+  private static schema = z.object({
+    order: z.object({
+      channel: z.object({
+        id: z.string(),
+        slug: z.string(),
+      }),
+      id: z.string(),
+      avataxId: z.string(),
+    }),
+    recipient: z.object({
+      privateMetadata: z.array(
+        z.object({
+          key: z.string(),
+          value: z.string(),
+        }),
+      ),
+    }),
+  });
+
+  private constructor(private data: z.infer<typeof SaleorCancelledOrderEvent.schema>) {}
+
+  getPrivateMetadata() {
+    return this.data.recipient.privateMetadata;
+  }
+
+  getChannelSlug() {
+    return this.data.order.channel.slug;
+  }
+
+  getAvataxId() {
+    return this.data.order.avataxId;
+  }
+
+  public static ParsingError = BaseError.subclass("AvataxAppSaleorOrderCancelledParsingError");
+
+  static create(payload: OrderCancelledPayloadFragment) {
+    if (!payload.order) {
```

Comment:
**Question:** I think zod gives us this by default so we don't need check 🤔

Reply:
Yes it does but I think it needs a little refactor and shouldn't be a part of this PR. We would have to basically improve the types for errors zod throws because they're incorrect (`mapErr` returns an array of errors and their contents don't reflect their type {at least in my testing}). 

If we want to have custom errors for, I guess, better Datadog visibility and different behaviours the solution would be to just replace these two checks with error mapping and then compare each with either `order` or `order.avataxId` and check what's the cause, then throw. To be honest I played around with it a bit and it doesn't seem like a great DX. Highlighted check is technically done twice but it's handling is way easier in my opinion. 

Alternatively we create [custom error mappings](https://zod.dev/ERROR_HANDLING?id=customizing-errors-with-zoderrormap) but still, we have to be very explicit and then still throw custom errors based on, I guess, message. 

I don't want to keep this PR blocked and I would explore this more on a tech-debt day.